### PR TITLE
Add GetDescriptorHandlerSingleMemory

### PR DESCRIPTION
**Pull request by twam**
GetDescriptorHandlerSingleMemory handles GET_DESCRIPTOR calls and
stores all descriptors in a single memory. It uses the same interface
as GetDescriptorHandler and is a drop-in replacement.

### CODE REVIEW

**Comment by zyp** [comment]
Here's some test results that I've posted on Discord before, but I'm adding them here as well for the record:

With the original `GetDescriptorHandler`:

```
Info: Device utilisation:
Info: 	       TRELLIS_SLICE:  5956/41820    14%
Info: 	              DP16KD:    21/  208    10%

Info: Max frequency for clock    '$glbnet$ulpi_clk$TRELLIS_IO_IN': 76.23 MHz (PASS at 60.00 MHz)
```

With the new `GetDescriptorHandlerSingleMemory`:

```
Info: Device utilisation:
Info: 	       TRELLIS_SLICE:  5497/41820    13%
Info: 	              DP16KD:    22/  208    10%

Info: Max frequency for clock    '$glbnet$ulpi_clk$TRELLIS_IO_IN': 96.98 MHz (PASS at 60.00 MHz)
```

My conclusion is that `GetDescriptorHandlerSingleMemory` saves a significant amount of logic and makes the critical path in the USB domain significantly shorter, at the cost of a single block ram.

It would be good to have a simple way of choosing which handler to use, but I'd be just as happy to simply see this become the new default handler.

File: luna/gateware/usb/usb2/descriptor_single_memory.py [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+
+        Each descriptor type starting from 0 until maximum used type number has
+        an entry consisting of number of indexes for this type number (2 bytes)
+        and address of first index (2 bytes).
+        Invalid entries have a value of 0x0000xxxx (0 entries).
+
+        Example:
+        0000  0xFFFF
+        0002  0xFFFF
+        0004  Number of device indexes
```
Comment (hansfbaier) [inline luna/gateware/usb/usb2/descriptor_single_memory.py:87]: Correct latin plural would be 'indices'

File: luna/gateware/usb/usb2/descriptor_single_memory.py [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+        0006  Address of first device index
+        0008  Number of configuration indexes
```
Comment (hansfbaier) [inline luna/gateware/usb/usb2/descriptor_single_memory.py:89]: idem :)

File: luna/gateware/usb/usb2/descriptor_single_memory.py [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+                yield
+            yield self.dut.tx.ready.eq(1)
+            yield
+
+        max_packet_length = 64
+        expected_data = raw_descriptor[start_position:]
+        expected_bytes = min(len(expected_data), max_length-start_position, max_packet_length)
+
+        for i in range(expected_bytes):
+            self.assertEqual((yield self.dut.tx.first), 1 if (i == 0) else 0)
```
Comment (hansfbaier) [inline luna/gateware/usb/usb2/descriptor_single_memory.py:362]: vertically aligning the second arguments might add a nice touch here

File: luna/gateware/usb/usb2/descriptor_single_memory.py [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+    @usb_domain_test_case
+    def test_unavailable_index_type(self):
+        # Unavailable index in between
+        yield from self._test_stall(0x10, 0, 0, 64)
+
+        # Index after last used type
+        yield from self._test_stall(0x42, 0, 0, 64)
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (hansfbaier) [inline luna/gateware/usb/usb2/descriptor_single_memory.py:438]: This is excellent, high quality work. Sorry for nitpicking a bit, but it is a by product of me reading the code. Some parts might be a matter of taste and conditioning, so feel free to do what you like anyway :)

File: luna/gateware/usb/usb2/descriptor_single_memory.py [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+                # If no entries are available for index or type number is unused, stall
+                with m.If(index >= rom_read_port.data.word_select(1, 16)):
+                    m.d.comb += [
+                        self.stall.eq(1)
+                    ]
+                    m.next = "IDLE"
+
+                with m.Else():
+                    m.d.comb += [
+                        rom_read_port.addr.eq(rom_read_port.data.bit_select(2, rom_read_port.addr.width)+index)
```
[the remark this answers is not in the document]
Comment (twam) [reply luna/gateware/usb/usb2/descriptor_single_memory.py:241]: I'm still unsure if this repo is following PEP-8 at all. My tooling shows quite a lot of warnings on all files and I don't know which to follow. :)

File: luna/gateware/usb/usb2/descriptor_single_memory.py (hunk 1 above)
Comment (twam) [reply luna/gateware/usb/usb2/descriptor_single_memory.py:87]: All sources I find (e.g. https://en.wiktionary.org/wiki/index) list `indexes` as well, often as their 'first' choice.

File: luna/gateware/usb/usb2/descriptor_single_memory.py [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+
+        # Write index offsets and data
+        for type_number, indexes in sorted(descriptors.items()):
+            for index, raw_descriptor in sorted(indexes.items()):
+                index_base_address = type_index_base_address[type_number] + index * 4
+                rom[index_base_address:index_base_address+4] = struct.pack(">HH", len(raw_descriptor), next_free_address)
+                rom[next_free_address:next_free_address+len(raw_descriptor)] = raw_descriptor
+                next_free_address += ((len(raw_descriptor)+3)//4)*4
+
+        # for i in range(len(rom)//16+1):
```
[the remark this answers is not in the document]
Comment (twam) [reply luna/gateware/usb/usb2/descriptor_single_memory.py:154]: Not needed, was only helpful for debugging. Will remove.

File: luna/gateware/usb/usb2/descriptor_single_memory.py [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+        on_first_packet = position_in_stream == self.start_position
+        on_last_packet = \
+            (position_in_stream == (descriptor_length - 1)) | \
+            (bytes_sent + 1 >= length)
+
+        with m.FSM() as fsm:
+            m.d.comb += self.tx.valid.eq(fsm.ongoing('STREAMING'))
+
+            with m.State('IDLE'):
+                m.d.comb += [
```
[the remark this answers is not in the document]
Comment (twam) [reply luna/gateware/usb/usb2/descriptor_single_memory.py:212]: Looks like most of the our luna code follows the one line approach. I therefore changed it accordingly.

File: luna/gateware/usb/usb2/descriptor_single_memory.py (hunk 3 above)
Comment (twam) [reply luna/gateware/usb/usb2/descriptor_single_memory.py:362]: Changed.

**Comment by twam** [comment]
@ktemkin I'd be happy to adapt this PR to either have it as a replacement for the 'old' GetDescriptorHandler or make it configurable. I can also move it into the `descriptor.py` and rename the class if desired. Just let me know, how you think it would be fitting best.

**Comment by ktemkin** [comment]
I'll figure out how to make this configurable and merge this next week. Thanks a lot. :)

**Comment by ktemkin** [comment]
Did a bit of cleanup, squished this into typical LUNA style, and made it the default when we're not avoiding block-RAMs. :)

See the final commit here:
https://github.com/greatscottgadgets/luna/commit/03607737813f56ba5c944e0d17559e27d8580055

Thanks a ton!

**Comment by twam** [comment]
Awesome! Thanks a lot.

**Pull request closed without merging**